File: backend/app/api/setting.py
"""系统设置接口：AI 模型（LLM）配置的读取 / 保存 / 连接测试。

**按用户隔离**：每个登录用户只能看到 / 修改自己的 AI 配置与供应商；
传别人的 provider_id 等同「不存在」（404），不会跨用户读写。

列表与设置回显只给脱敏预览；仅「编辑某条自己的供应商」时可按 id 单独取回明文
Key（见 /llm/providers/{id}/api-key），用于直接修改。
"""
from typing import Annotated

from fastapi import APIRouter, Depends, Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.core.database import get_db
from app.core.runtime_config import LLMConfig
from app.models.user import User
from app.schemas.setting import (
    LLMApiKeyOut,
    LLMFetchModelsRequest,
    LLMFetchModelsResult,
    LLMModelsOut,
    LLMProviderCreate,
    LLMProviderOut,
    LLMProviderReorder,
    LLMProviderUpdate,
    LLMSettingOut,
    LLMSettingUpdate,
    LLMTestRequest,
    LLMTestResult,
)
from app.services import settings as settings_service

router = APIRouter(prefix="/api/settings", tags=["settings"])


def _to_out(cfg: LLMConfig, source: str) -> LLMSettingOut:
    mode = "real" if cfg.ready else "mock"
    if mode == "real":
        message = f"已启用真实模型（{cfg.model}）"
    elif cfg.enabled and not cfg.api_key:
        message = "已开启 AI 分析，但尚未填写 API Key，当前仍走规则 Mock"
    elif cfg.enabled:
        message = "已开启 AI 分析，但配置不完整（缺少地址或模型），当前仍走规则 Mock"
    else:
        message = "未启用 AI 分析，当前走规则 Mock（无需 Key 也能跑通全链路）"

    return LLMSettingOut(
        enabled=cfg.enabled,
        has_api_key=bool(cfg.api_key),
        api_key_preview=settings_service.mask_api_key(cfg.api_key),
        base_url=cfg.base_url,
        model=cfg.model,
        timeout_seconds=cfg.timeout_seconds,
        min_change_lines=cfg.min_change_lines,
        mode=mode,
        source=source,
        message=message,
    )


@router.get("/llm", response_model=LLMSettingOut)
async def read_llm_setting(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """读取当前账号的 AI 模型配置（Key 脱敏）。"""
    cfg, source = await settings_service.get_llm_setting(db, current_user.id)
    return _to_out(cfg, source)


@router.put("/llm", response_model=LLMSettingOut)
async def update_llm_setting(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    payload: LLMSettingUpdate,
):
    """保存当前账号的 AI 模型配置，保存后立即生效、无需重启。"""
    cfg, source = await settings_service.save_llm_setting(db, current_user.id, payload)
    return _to_out(cfg, source)


@router.post("/llm/test", response_model=LLMTestResult)
async def test_llm_setting(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    payload: LLMTestRequest,
):
    """测试模型连通性：用界面填的值（缺省用自己已保存的）发一次最小请求。"""
    return await settings_service.test_llm_connection(db, current_user.id, payload)


# ---- 供应商管理：多个模型服务商，运行时使用「使用中」的那一条 ----


@router.get("/llm/providers", response_model=list[LLMProviderOut])
async def list_llm_providers(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """当前账号已配置的供应商列表（Key 脱敏）。"""
    return await settings_service.list_providers(db, current_user.id)


@router.post("/llm/providers", response_model=LLMProviderOut)
async def create_llm_provider(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    payload: LLMProviderCreate,
):
    """添加供应商；当前账号的第一条自动设为「使用中」。"""
    return await settings_service.create_provider(db, current_user.id, payload)


@router.post("/llm/providers/models", response_model=LLMFetchModelsResult)
async def fetch_llm_provider_models(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    payload: LLMFetchModelsRequest,
):
    """从上游供应商拉取模型列表（OpenAI 兼容的 GET /models），用于界面下拉选择。"""
    return await settings_service.fetch_provider_models(db, current_user.id, payload)


@router.post("/llm/providers/reorder", response_model=None)
async def reorder_llm_providers(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    payload: LLMProviderReorder,
):
    """拖动排序：按期望顺序传一组自己的供应商 id（仅改展示顺序）。"""
    await settings_service.reorder_providers(db, current_user.id, payload.ids)


@router.put("/llm/providers/{provider_id}", response_model=LLMProviderOut)
async def update_llm_provider(
    provider_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    payload: LLMProviderUpdate,
):
    """编辑自己的供应商（只改传了的字段；别人的 id 返回 404）。"""
    return await settings_service.update_provider(
        db, current_user.id, provider_id, payload
    )


@router.delete("/llm/providers/{provider_id}", response_model=None)
async def delete_llm_provider(
    provider_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """删除自己的供应商；删的是「使用中」的则自动切换到最新一条。"""
    await settings_service.delete_provider(db, current_user.id, provider_id)


@router.get("/llm/providers/{provider_id}/models", response_model=LLMModelsOut)
async def read_llm_provider_models(
    provider_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """读取自己某条供应商已缓存的完整模型列表（列表接口只回数量）。"""
    return LLMModelsOut(
        models=await settings_service.get_provider_models(
            db, current_user.id, provider_id
        )
    )


@router.get("/llm/providers/{provider_id}/api-key", response_model=LLMApiKeyOut)
async def read_llm_provider_api_key(
    provider_id: int,
    response: Response,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """读取自己某条供应商已保存的 API Key 明文（编辑弹窗回显用）。

    明文只用于界面查看，明确禁止任何缓存，避免留存在浏览器缓存/磁盘里；
    且只有该供应商的归属用户能取到。
    """
    response.headers["Cache-Control"] = "no-store"
    response.headers["Pragma"] = "no-cache"
    return LLMApiKeyOut(
        api_key=await settings_service.get_provider_api_key(
            db, current_user.id, provider_id
        )
    )


@router.post("/llm/providers/{provider_id}/activate", response_model=LLMProviderOut)
async def activate_llm_provider(
    provider_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """切换自己「使用中」的供应商。"""
    return await settings_service.activate_provider(
        db, current_user.id, provider_id
    )


@router.post("/llm/providers/{provider_id}/duplicate", response_model=LLMProviderOut)
async def duplicate_llm_provider(
    provider_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """复制自己的供应商（含已保存的 Key），副本追加到末尾且不设为「使用中」。"""
    return await settings_service.duplicate_provider(
        db, current_user.id, provider_id
    )

File: frontend/src/api/setting.ts
import request from "@/api/request";

/** 当前 AI 模型配置（Key 已脱敏，不含明文） */
export interface LlmSetting {
  enabled: boolean;
  hasApiKey: boolean;
  /** 脱敏预览，如 sk-****abcd */
  apiKeyPreview: string;
  baseUrl: string;
  model: string;
  timeoutSeconds: number;
  minChangeLines: number;
  /** real = 走真实模型；mock = 规则兜底 */
  mode: "real" | "mock";
  /** database = 已在设置页保存；env = 读的是 .env 默认值 */
  source: "database" | "env";
  message: string;
}

/** 保存设置的入参（只提交要改的字段） */
export interface LlmSettingUpdate {
  enabled?: boolean;
  /** 留空表示不修改；有值则覆盖 */
  apiKey?: string;
  /** true 时无条件清除已保存的 Key */
  clearApiKey?: boolean;
  baseUrl?: string;
  model?: string;
  timeoutSeconds?: number;
  minChangeLines?: number;
}

export interface LlmTestRequest {
  /** 测试已保存的供应商时传入：它的 Key 前端拿不到（脱敏），由后端自己取 */
  providerId?: number;
  baseUrl?: string;
  model?: string;
  apiKey?: string;
  timeoutSeconds?: number;
}

export interface LlmTestResult {
  ok: boolean;
  message: string;
  model: string;
  latencyMs: number;
}

/** 读取当前 AI 模型配置 */
export function fetchLlmSetting(): Promise<LlmSetting> {
  return request.get<unknown, LlmSetting>("/api/settings/llm");
}

/** 保存 AI 模型配置（保存后立即生效） */
export function updateLlmSetting(payload: LlmSettingUpdate): Promise<LlmSetting> {
  return request.put<unknown, LlmSetting>("/api/settings/llm", payload);
}

/**
 * 测试模型连通性。
 *
 * 后端按 timeoutSeconds 等待模型返回，前端必须比它更晚超时，否则模型还没回、
 * 前端就先报 "timeout of 10000ms exceeded"。这里取请求里的 timeoutSeconds + 5s 缓冲。
 */
export function testLlmSetting(payload: LlmTestRequest): Promise<LlmTestResult> {
  const waitMs = ((payload.timeoutSeconds ?? 30) + 5) * 1000;
  return request.post<unknown, LlmTestResult>(
    "/api/settings/llm/test",
    payload,
    { timeout: waitMs },
  );
}

/** 已配置的模型供应商（Key 已脱敏，不含明文） */
export interface LlmProvider {
  id: number;
  /** 展示名，如 DeepSeek；空串表示未命名（界面显示「自定义」） */
  name: string;
  baseUrl: string;
  model: string;
  hasApiKey: boolean;
  apiKeyPreview: string;
  /** 运行时正在使用这一条 */
  isActive: boolean;
  /** 已缓存的模型数量；完整列表用 fetchLlmProviderModels 按需取 */
  modelsCount: number;
}

/** 添加供应商的入参 */
export interface LlmProviderCreatePayload {
  name?: string;
  baseUrl: string;
  model: string;
  apiKey: string;
  /** 添加前已从上游拉到的模型列表（可选） */
  models?: string[];
}

/** 编辑供应商的入参（只提交要改的字段） */
export interface LlmProviderUpdatePayload {
  name?: string;
  baseUrl?: string;
  model?: string;
  /** 留空表示不修改；有值则覆盖 */
  apiKey?: string;
  clearApiKey?: boolean;
  /** 省略 = 保持不变；传了就覆盖（保存本次新拉到的模型列表） */
  models?: string[];
}

/** 已配置的供应商列表（使用中的排最前） */
export function fetchLlmProviders(): Promise<LlmProvider[]> {
  return request.get<unknown, LlmProvider[]>("/api/settings/llm/providers");
}

/** 添加供应商（第一条自动设为「使用中」并立即生效） */
export function createLlmProvider(payload: LlmProviderCreatePayload): Promise<LlmProvider> {
  return request.post<unknown, LlmProvider>("/api/settings/llm/providers", payload);
}

/** 编辑供应商（只改传了的字段） */
export function updateLlmProvider(
  id: number,
  payload: LlmProviderUpdatePayload,
): Promise<LlmProvider> {
  return request.put<unknown, LlmProvider>(`/api/settings/llm/providers/${id}`, payload);
}

/** 删除供应商；删的是「使用中」的则自动切换到最新一条 */
export function deleteLlmProvider(id: number): Promise<void> {
  return request.delete(`/api/settings/llm/providers/${id}`);
}

/** 拖动排序：按期望顺序传一组供应商 id（仅改展示顺序） */
export function reorderLlmProviders(ids: number[]): Promise<void> {
  return request.post<unknown, void>("/api/settings/llm/providers/reorder", { ids });
}

/** 切换「使用中」的供应商，运行时立即改用它的配置 */
export function activateLlmProvider(id: number): Promise<LlmProvider> {
  return request.post<unknown, LlmProvider>(`/api/settings/llm/providers/${id}/activate`);
}

/** 复制供应商（含已保存的 Key），副本追加到末尾且不设为「使用中」 */
export function duplicateLlmProvider(id: number): Promise<LlmProvider> {
  return request.post<unknown, LlmProvider>(`/api/settings/llm/providers/${id}/duplicate`);
}

/** 读取某条已保存供应商的 API Key 明文（编辑弹窗回显用；列表只给脱敏值） */
export function fetchLlmProviderApiKey(id: number): Promise<string> {
  return request
    .get<unknown, { apiKey: string }>(`/api/settings/llm/providers/${id}/api-key`)
    .then((res) => res.apiKey || "");
}

/** 从上游供应商拉取模型列表的入参 */
export interface LlmFetchModelsRequest {
  /** 测试已保存的供应商时传入：Key 前端拿不到（脱敏），由后端自己取 */
  providerId?: number;
  baseUrl?: string;
  apiKey?: string;
  timeoutSeconds?: number;
}

/** 从上游供应商拉取模型列表的结果 */
export interface LlmFetchModelsResult {
  ok: boolean;
  models: string[];
  message: string;
}

/**
 * 从上游供应商拉取模型列表（OpenAI 兼容的 GET /models）。
 * 后端按 timeoutSeconds 等待返回，前端必须比它更晚超时。
 */
export function fetchProviderModels(
  payload: LlmFetchModelsRequest,
): Promise<LlmFetchModelsResult> {
  const waitMs = ((payload.timeoutSeconds ?? 15) + 5) * 1000;
  return request.post<unknown, LlmFetchModelsResult>(
    "/api/settings/llm/providers/models",
    payload,
    { timeout: waitMs },
  );
}

/** 读取某条已保存供应商缓存的完整模型列表（列表接口只回数量） */
export function fetchLlmProviderModels(id: number): Promise<string[]> {
  return request
    .get<unknown, { models: string[] }>(`/api/settings/llm/providers/${id}/models`)
    .then((res) => res.models || []);
}
